rfcs/0137-nix-language-version.md

```diff
@@ -0,0 +1,113 @@
+---
+feature: nix-language-version
+start-date: 2022-12-12
+author: @fricklerhandwerk
+co-authors: @thufschmitt @Ericson2314
+shepherd-team: 
+shepherd-leader: 
+related-issues: https://github.com/NixOS/nix/issues/7255
+---
+
+# Summary
+[summary]: #summary
+
+Introduce a convention to determine which version of the Nix language grammar to use for parsing a Nix file.
+
+# Motivation
+[motivation]: #motivation
+
+Currently it's impossible to introduce backwards-incompatible changes to the Nix language without breaking existing setups.
+This proposal is first step towards overcoming that limitation.
+
+# Detailed design
+[design]: #detailed-design
+
+Introduce a magic comment in the first line of a Nix file:
+
+    #? Nix <version>
+
+where `<version>` is a released version of Nix the given file is intended to work with.
```

Comment:
What should the default assumed value be? There's some options:
- Hardcode the default to the current Nix version `2.12.0`, this way we always know what the field is if not set, and if you use newer Nix language features you need to opt-in to use them. But this then begs the question, how do we migrate to newer defaults? This would lead to requiring the field to be mandatory at some point.
- The Nix version currently evaluating the file, but this has the problem that if a user writes a file with newer Nix version and somebody with an older Nix version tries to evaluate it, it fails without any good error message

Or should the field be mandatory with a warning (and one of the above defaults), or an error? If so, Nix could then also analyze the file to figure out which versions it supports and suggest using the oldest one of those. This then probably only works well with syntactic changes though, not semantic ones. It's also very complex to implement, doesn't really allow deprecating things at all.

Reply:
Alternatively, if the Nix version was given in the `flake.nix` instead, it would be much more realistic to always require setting the field, since there's flake templates that can be evolved. Also a flake sets the version for the entire Flake, not just a single Nix file, so there's no big overhead to changing/assigning it. Or we could even combine the two: Allow setting the default Nix (language) version in the `flake.nix`, but allow overriding it in individual Nix files

Reply:
I think the current version should be assumed. At some point we can consider a warning if you are on a "very old" version (including if you haven't specified one).

Reply:
```suggestion
where `<version>` is a released version of Nix the given file is intended to work with.

If no magic comment is present, Nix 2.12 – the stable release at the time this RFC was accepted – is assumed.
```

Sounds reasonable.